connmgr/connmanager.go

```diff
@@ -530,8 +525,8 @@ func (cm *ConnManager) Connect(ctx context.Context, c *ConnReq) {
 		return
 	}
 
-	if atomic.LoadUint64(&c.id) == 0 {
-		atomic.StoreUint64(&c.id, atomic.AddUint64(&cm.connReqCount, 1))
+	if c.id.Load() == 0 {
```

Comment:
Copying @jrick's comments from #3053:

> not the fault of your patch, but this looks very racy. c.id may not longer be 0 by the time we store the value.

> I don't see a straightforward way to implement this correctly using atomics. Both c.id and cm.connReqCount need to be updated atomically together. This might require a mutex instead.

Reply:
I'll considering addressing this by adding a mutex just to be extra certain, but I don't think this is really an issue after looking at it more closely.  It would most definitely be an issue if the ID were be assigned anywhere else concurrently, but this is the only place an **existing** connection request has an ID assigned and so hitting a race here would require calling `Connect` on the same connection request concurrently, which doesn't happen.

The only other case of an ID being assigned is in `newConnReq`, but as the name indicates, that creates a _new_ connection request and assigned the ID at initialization time, so there is no possible race with this.

Reply:
Done in #3055.